File: goldenflow/transforms/text.py
from __future__ import annotations

import re
import unicodedata

import polars as pl

from goldenflow.transforms import register_transform


@register_transform(name="strip", input_types=["string"], auto_apply=True, priority=90, mode="expr")
def strip(column: str) -> pl.Expr:
    return pl.col(column).str.strip_chars()


@register_transform(
    name="lowercase", input_types=["string"], auto_apply=False, priority=50, mode="expr"
)
def lowercase(column: str) -> pl.Expr:
    return pl.col(column).str.to_lowercase()


@register_transform(
    name="uppercase", input_types=["string"], auto_apply=False, priority=50, mode="expr"
)
def uppercase(column: str) -> pl.Expr:
    return pl.col(column).str.to_uppercase()


@register_transform(
    name="title_case", input_types=["string"], auto_apply=False, priority=50, mode="expr"
)
def title_case(column: str) -> pl.Expr:
    return pl.col(column).str.to_titlecase()


@register_transform(
    name="normalize_unicode", input_types=["string"], auto_apply=True, priority=85, mode="series"
)
def normalize_unicode(series: pl.Series) -> pl.Series:
    def _normalize(val: str | None) -> str | None:
        if val is None:
            return None
        nfkd = unicodedata.normalize("NFKD", val)
        return "".join(c for c in nfkd if not unicodedata.combining(c))

    return series.map_elements(_normalize, return_dtype=pl.Utf8)


@register_transform(
    name="remove_punctuation", input_types=["string"], auto_apply=False, priority=40, mode="expr"
)
def remove_punctuation(column: str) -> pl.Expr:
    return pl.col(column).str.replace_all(r"[^a-zA-Z0-9\s]", "")


@register_transform(
    name="collapse_whitespace", input_types=["string"], auto_apply=True, priority=80, mode="expr"
)
def collapse_whitespace(column: str) -> pl.Expr:
    return pl.col(column).str.replace_all(r"\s{2,}", " ")


@register_transform(
    name="truncate", input_types=["string"], auto_apply=False, priority=30, mode="series"
)
def truncate(series: pl.Series, n: int = 255) -> pl.Series:
    return series.str.slice(0, n)


@register_transform(
    name="normalize_quotes",
    input_types=["string"],
    auto_apply=True,
    priority=84,
    mode="expr",
)
def normalize_quotes(column: str) -> pl.Expr:
    """Replace smart/curly quotes with straight quotes."""
    return (
        pl.col(column)
        .str.replace_all("\u201c", '"')   # left double
        .str.replace_all("\u201d", '"')   # right double
        .str.replace_all("\u2018", "'")   # left single
        .str.replace_all("\u2019", "'")   # right single
        .str.replace_all("\u2033", '"')   # double prime
        .str.replace_all("\u2032", "'")   # prime
    )


@register_transform(
    name="remove_html_tags",
    input_types=["string"],
    auto_apply=False,
    priority=45,
    mode="series",
)
def remove_html_tags(series: pl.Series) -> pl.Series:
    """Strip HTML tags from text."""
    _tag_re = re.compile(r"<[^>]+>")

    def _strip(val: str | None) -> str | None:
        if val is None:
            return None
        return _tag_re.sub("", val)

    return series.map_elements(_strip, return_dtype=pl.Utf8)


@register_transform(
    name="remove_urls",
    input_types=["string"],
    auto_apply=False,
    priority=40,
    mode="series",
)
def remove_urls(series: pl.Series) -> pl.Series:
    """Strip URLs (http/https) from text."""
    _url_re = re.compile(r"https?://\S+")

    def _strip(val: str | None) -> str | None:
        if val is None:
            return None
        return _url_re.sub("", val)

    return series.map_elements(_strip, return_dtype=pl.Utf8)


@register_transform(
    name="remove_digits",
    input_types=["string"],
    auto_apply=False,
    priority=35,
    mode="expr",
)
def remove_digits(column: str) -> pl.Expr:
    """Remove all digit characters from text."""
    return pl.col(column).str.replace_all(r"\d", "")


@register_transform(
    name="pad_left",
    input_types=["string"],
    auto_apply=False,
    priority=30,
    mode="series",
)
def pad_left(series: pl.Series, width: int = 10, char: str = "0") -> pl.Series:
    """Left-pad strings to a fixed width."""

    def _pad(val: str | None) -> str | None:
        if val is None:
            return None
        return val.rjust(width, char)

    return series.map_elements(_pad, return_dtype=pl.Utf8)


@register_transform(
    name="pad_right",
    input_types=["string"],
    auto_apply=False,
    priority=30,
    mode="series",
)
def pad_right(series: pl.Series, width: int = 10, char: str = " ") -> pl.Series:
    """Right-pad strings to a fixed width."""

    def _pad(val: str | None) -> str | None:
        if val is None:
            return None
        return val.ljust(width, char)

    return series.map_elements(_pad, return_dtype=pl.Utf8)


_EMOJI_RE = re.compile(
    "["
    "\U0001f600-\U0001f64f"  # emoticons
    "\U0001f300-\U0001f5ff"  # symbols & pictographs
    "\U0001f680-\U0001f6ff"  # transport & map
    "\U0001f1e0-\U0001f1ff"  # flags
    "\U00002702-\U000027b0"  # dingbats
    "\U000024c2-\U0001f251"  # enclosed characters
    "\U0001f900-\U0001f9ff"  # supplemental symbols
    "\U0001fa00-\U0001fa6f"  # chess symbols
    "\U0001fa70-\U0001faff"  # symbols extended-A
    "\U00002600-\U000026ff"  # misc symbols
    "\U0000200d"             # zero-width joiner
    "\U0000fe0f"             # variation selector
    "]+",
    flags=re.UNICODE,
)


@register_transform(
    name="remove_emojis",
    input_types=["string"],
    auto_apply=False,
    priority=38,
    mode="series",
)
def remove_emojis(series: pl.Series) -> pl.Series:
    """Remove emoji characters from text."""

    def _strip(val: str | None) -> str | None:
        if val is None:
            return None
        return _EMOJI_RE.sub("", val)

    return series.map_elements(_strip, return_dtype=pl.Utf8)


@register_transform(
    name="fix_mojibake",
    input_types=["string"],
    auto_apply=False,
    priority=86,
    mode="series",
)
def fix_mojibake(series: pl.Series) -> pl.Series:
    """Fix common UTF-8/Latin-1 mojibake by re-encoding."""

    def _fix(val: str | None) -> str | None:
        if val is None:
            return None
        try:
            return val.encode("latin-1").decode("utf-8")
        except (UnicodeDecodeError, UnicodeEncodeError):
            return val

    return series.map_elements(_fix, return_dtype=pl.Utf8)


@register_transform(
    name="normalize_line_endings",
    input_types=["string"],
    auto_apply=False,
    priority=82,
    mode="series",
)
def normalize_line_endings(series: pl.Series) -> pl.Series:
    """Normalize \\r\\n and \\r to \\n."""

    def _norm(val: str | None) -> str | None:
        if val is None:
            return None
        return val.replace("\r\n", "\n").replace("\r", "\n")

    return series.map_elements(_norm, return_dtype=pl.Utf8)


_NUMBER_RE = re.compile(r"\d+\.?\d*")


@register_transform(
    name="extract_numbers",
    input_types=["string"],
    auto_apply=False,
    priority=30,
    mode="series",
)
def extract_numbers(series: pl.Series) -> pl.Series:
    """Extract all numbers from text, joined by spaces."""

    def _extract(val: str | None) -> str | None:
        if val is None:
            return None
        return " ".join(_NUMBER_RE.findall(val))

    return series.map_elements(_extract, return_dtype=pl.Utf8)

File: packages/goldenflow-js/src/core/transforms/text.ts
/**
 * Text transforms — ported from goldenflow/transforms/text.py
 * Side-effect module: registers 18 text transforms on import.
 */

import type { ColumnValue } from "../types.js";
import { registerTransform } from "./registry.js";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function mapStrings(
  values: readonly ColumnValue[],
  fn: (s: string) => string,
): ColumnValue[] {
  return values.map((v) => {
    if (v === null || typeof v !== "string") return v;
    return fn(v);
  });
}

// ---------------------------------------------------------------------------
// strip (priority 90, auto_apply, string)
// ---------------------------------------------------------------------------

function strip(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.trim());
}

registerTransform(
  { name: "strip", inputTypes: ["string"], autoApply: true, priority: 90, mode: "expr" },
  strip,
);

// ---------------------------------------------------------------------------
// lowercase (50, string)
// ---------------------------------------------------------------------------

function lowercase(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.toLowerCase());
}

registerTransform(
  { name: "lowercase", inputTypes: ["string"], priority: 50, mode: "expr" },
  lowercase,
);

// ---------------------------------------------------------------------------
// uppercase (50, string)
// ---------------------------------------------------------------------------

function uppercase(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.toUpperCase());
}

registerTransform(
  { name: "uppercase", inputTypes: ["string"], priority: 50, mode: "expr" },
  uppercase,
);

// ---------------------------------------------------------------------------
// title_case (50, string)
// ---------------------------------------------------------------------------

function titleCase(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) =>
    s.toLowerCase().replace(/\b\w/g, (ch) => ch.toUpperCase()),
  );
}

registerTransform(
  { name: "title_case", inputTypes: ["string"], priority: 50, mode: "expr" },
  titleCase,
);

// ---------------------------------------------------------------------------
// normalize_unicode (85, auto_apply, string)
// ---------------------------------------------------------------------------

function normalizeUnicode(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) =>
    s.normalize("NFKD").replace(/\p{M}/gu, ""),
  );
}

registerTransform(
  { name: "normalize_unicode", inputTypes: ["string"], autoApply: true, priority: 85, mode: "series" },
  normalizeUnicode,
);

// ---------------------------------------------------------------------------
// remove_punctuation (40, string)
// ---------------------------------------------------------------------------

function removePunctuation(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.replace(/[^\w\s]/g, ""));
}

registerTransform(
  { name: "remove_punctuation", inputTypes: ["string"], priority: 40, mode: "series" },
  removePunctuation,
);

// ---------------------------------------------------------------------------
// collapse_whitespace (80, auto_apply, string)
// ---------------------------------------------------------------------------

function collapseWhitespace(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.replace(/\s+/g, " ").trim());
}

registerTransform(
  { name: "collapse_whitespace", inputTypes: ["string"], autoApply: true, priority: 80, mode: "expr" },
  collapseWhitespace,
);

// ---------------------------------------------------------------------------
// truncate (30, string, param: n=255)
// ---------------------------------------------------------------------------

function truncate(values: readonly ColumnValue[], n: unknown = 255): ColumnValue[] {
  const maxLen = typeof n === "number" ? n : Number(n) || 255;
  return mapStrings(values, (s) => s.slice(0, maxLen));
}

registerTransform(
  { name: "truncate", inputTypes: ["string"], priority: 30, mode: "series" },
  truncate,
);

// ---------------------------------------------------------------------------
// normalize_quotes (84, auto_apply, string)
// ---------------------------------------------------------------------------

function normalizeQuotes(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) =>
    s
      .replace(/[\u2018\u2019\u201A\u201B]/g, "'")
      .replace(/[\u201C\u201D\u201E\u201F]/g, '"'),
  );
}

registerTransform(
  { name: "normalize_quotes", inputTypes: ["string"], autoApply: true, priority: 84, mode: "series" },
  normalizeQuotes,
);

// ---------------------------------------------------------------------------
// remove_html_tags (45, string)
// ---------------------------------------------------------------------------

function removeHtmlTags(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.replace(/<[^>]*>/g, ""));
}

registerTransform(
  { name: "remove_html_tags", inputTypes: ["string"], priority: 45, mode: "series" },
  removeHtmlTags,
);

// ---------------------------------------------------------------------------
// remove_urls (40, string)
// ---------------------------------------------------------------------------

function removeUrls(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) =>
    s.replace(/https?:\/\/[^\s]+/g, "").trim(),
  );
}

registerTransform(
  { name: "remove_urls", inputTypes: ["string"], priority: 40, mode: "series" },
  removeUrls,
);

// ---------------------------------------------------------------------------
// remove_digits (35, string)
// ---------------------------------------------------------------------------

function removeDigits(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.replace(/\d/g, ""));
}

registerTransform(
  { name: "remove_digits", inputTypes: ["string"], priority: 35, mode: "series" },
  removeDigits,
);

// ---------------------------------------------------------------------------
// pad_left (30, string, params: width=10, char="0")
// ---------------------------------------------------------------------------

function padLeft(
  values: readonly ColumnValue[],
  width: unknown = 10,
  char: unknown = "0",
): ColumnValue[] {
  const w = typeof width === "number" ? width : Number(width) || 10;
  const c = typeof char === "string" ? char : "0";
  return mapStrings(values, (s) => s.padStart(w, c));
}

registerTransform(
  { name: "pad_left", inputTypes: ["string"], priority: 30, mode: "series" },
  padLeft,
);

// ---------------------------------------------------------------------------
// pad_right (30, string, params: width=10, char=" ")
// ---------------------------------------------------------------------------

function padRight(
  values: readonly ColumnValue[],
  width: unknown = 10,
  char: unknown = " ",
): ColumnValue[] {
  const w = typeof width === "number" ? width : Number(width) || 10;
  const c = typeof char === "string" ? char : " ";
  return mapStrings(values, (s) => s.padEnd(w, c));
}

registerTransform(
  { name: "pad_right", inputTypes: ["string"], priority: 30, mode: "series" },
  padRight,
);

// ---------------------------------------------------------------------------
// remove_emojis (38, string)
// ---------------------------------------------------------------------------

function removeEmojis(values: readonly ColumnValue[]): ColumnValue[] {
  // Covers common emoji Unicode ranges including emoticons, symbols, dingbats,
  // supplemental symbols, flags, and extended pictographic.
  const emojiPattern =
    /[\u{1F600}-\u{1F64F}\u{1F300}-\u{1F5FF}\u{1F680}-\u{1F6FF}\u{1F1E0}-\u{1F1FF}\u{2600}-\u{26FF}\u{2700}-\u{27BF}\u{FE00}-\u{FE0F}\u{1F900}-\u{1F9FF}\u{1FA00}-\u{1FA6F}\u{1FA70}-\u{1FAFF}\u{200D}\u{20E3}\u{E0020}-\u{E007F}]/gu;
  return mapStrings(values, (s) => s.replace(emojiPattern, ""));
}

registerTransform(
  { name: "remove_emojis", inputTypes: ["string"], priority: 38, mode: "series" },
  removeEmojis,
);

// ---------------------------------------------------------------------------
// fix_mojibake (86, string)
// ---------------------------------------------------------------------------

function fixMojibake(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => {
    try {
      // Attempt latin1 -> utf8 re-encoding: encode string bytes as latin1,
      // then decode as UTF-8. If the result is valid and different, use it.
      const encoder = new TextEncoder();
      const bytes = new Uint8Array(s.length);
      for (let i = 0; i < s.length; i++) {
        const code = s.charCodeAt(i);
        if (code > 255) return s; // Not latin1-encodable; skip
        bytes[i] = code;
      }
      const decoded = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
      return decoded;
    } catch {
      // If decoding fails the string is not mojibake; return as-is
      return s;
    }
  });
}

registerTransform(
  { name: "fix_mojibake", inputTypes: ["string"], priority: 86, mode: "series" },
  fixMojibake,
);

// ---------------------------------------------------------------------------
// normalize_line_endings (82, string)
// ---------------------------------------------------------------------------

function normalizeLineEndings(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => s.replace(/\r\n/g, "\n").replace(/\r/g, "\n"));
}

registerTransform(
  { name: "normalize_line_endings", inputTypes: ["string"], priority: 82, mode: "series" },
  normalizeLineEndings,
);

// ---------------------------------------------------------------------------
// extract_numbers (30, string)
// ---------------------------------------------------------------------------

function extractNumbers(values: readonly ColumnValue[]): ColumnValue[] {
  return mapStrings(values, (s) => {
    const nums = s.match(/-?\d+(?:\.\d+)?/g);
    return nums ? nums.join(" ") : "";
  });
}

registerTransform(
  { name: "extract_numbers", inputTypes: ["string"], priority: 30, mode: "series" },
  extractNumbers,
);
